cmd/kcp-front-proxy/main.go

```
@@ -95,23 +98,41 @@ routed based on paths.`,
 				return err
 			}
 
-			// start index
-			rootConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&clientcmd.ClientConfigLoadingRules{ExplicitPath: options.RootKubeconfig}, nil).ClientConfig()
+			// get root API identities
+			nonIdentityRootConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&clientcmd.ClientConfigLoadingRules{ExplicitPath: options.RootKubeconfig}, nil).ClientConfig()
 			if err != nil {
 				return fmt.Errorf("failed to load root kubeconfig: %w", err)
 			}
-			rootShardConfig, err := kcpclient.NewClusterForConfig(rootConfig)
+			nonIdentityRootKcpClusterClient, err := kcpclient.NewClusterForConfig(nonIdentityRootConfig)
 			if err != nil {
 				return fmt.Errorf("failed to create root client: %w", err)
 			}
-			kcpSharedInformerFactory := kcpinformers.NewSharedInformerFactoryWithOptions(rootShardConfig.Cluster(tenancyv1alpha1.RootCluster), 30*time.Minute)
+			rootShardConfig, resolveIdentities := bootstrap.NewConfigWithWildcardIdentities(nonIdentityRootConfig, bootstrap.KcpRootGroupExportNames, bootstrap.KcpRootGroupResourceExportNames, nonIdentityRootKcpClusterClient.Cluster(tenancyv1alpha1.RootCluster))
+			if err := wait.PollImmediateInfiniteWithContext(ctx, time.Millisecond*500, func(ctx context.Context) (bool, error) {
+				if err := resolveIdentities(ctx); err != nil {
+					klog.V(3).Infof("failed to resolve identities, keeping trying: %v", err)
+					return false, nil
+				}
+				return true, nil
+			}); err != nil {
+				klog.Errorf("failed to get or create identities: %v", err)
+				// nolint:nilerr
+				return nil // don't klog.Fatal. This only happens when context is cancelled.
```

I feel like we could return the error here?